gk/fib.c

```diff
@@ -0,0 +1,1440 @@
+			return NULL;
+
+		neigh_fib = &ltbl->fib_tbl6[fib_id];
+
+		/*
+		 * Invalid gateway entry, since the neighbor entry
+		 * and the gateway entry should be in the same network.
+		 */
+		if ((action == GK_FWD_GATEWAY_FRONT_NET &&
+				neigh_fib->action != GK_FWD_NEIGHBOR_FRONT_NET)
+				|| (action == GK_FWD_GATEWAY_BACK_NET &&
+				neigh_fib->action !=
+				GK_FWD_NEIGHBOR_BACK_NET)) {
+			return NULL;
+		}
+
+		return neigh_fib;
+	} else {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: unconfigued IP type %hu at interface %s!\n",
+			gw_addr->proto, iface->name);
+	}
+
+	return NULL;
+}
+
+/*
+ * This function is called by del_fib_entry_locked().
+ * Notice that, it doesn't stand on its own, and it's only
+ * a construct to make del_fib_entry_locked() readable.
+ */
+static int
+del_gateway_from_neigh_table_locked(
+	struct ip_prefix *ip_prefix, enum gk_fib_action action,
+	struct ether_cache *eth_cache, struct gk_config *gk_conf)
+{
+	int ret;
+	struct gatekeeper_if *iface;
+	struct ipaddr *gw_addr = &eth_cache->ip_addr;
+	struct gk_fib *neigh_fib;
+
+	if (action == GK_FWD_GATEWAY_FRONT_NET)
+		iface = &gk_conf->net->front;
+	else if (likely(action == GK_FWD_GATEWAY_BACK_NET))
+		iface = &gk_conf->net->back;
+	else {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to delete a Gateway ethernet cache entry from neighbor hash table, since it has invalid action %d.\n",
+			action);
+		return -1;
+	}
+
+	/*
+	 * Find the FIB entry for the gateway.
+	 * We need to release the @eth_cache
+	 * Ethernet header entry from the neighbor hash table.
+	 */
+	neigh_fib = find_fib_entry_for_neighbor_locked(
+			gw_addr, action, iface, gk_conf);
+	if (neigh_fib == NULL)
+		return -1;
+
+	ret = ether_cache_put(neigh_fib, eth_cache, gw_addr);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to release the Ethernet cached header of the Grantor FIB entry for the IP prefix %s at %s\n",
+			ip_prefix->str, __func__);
+		return -1;
+	}
+
+	return 0;
+}
+
+/*
+ * For removing FIB entries, once the field @ref_cnt
+ * of a FIB entry becomes 0, then it needs to notify the GK instances
+ * to flush the flows that have a reference to it. This can be
+ * implemented by mailbox ring.
+ */
+static int
+del_fib_entry_locked(struct ip_prefix *ip_prefix, struct gk_config *gk_conf)
+{
+	int ret;
+	struct gk_fib *ip_prefix_fib;
+
+	ip_prefix_fib = check_fib_entry_for_del_locked(
+		ip_prefix, &ip_prefix->addr, ip_prefix->len, gk_conf);
+	if (ip_prefix_fib == NULL)
+		return -1;
+
+	/*
+	 * We need to notify the GK blocks whenever we remove
+	 * a FIB entry that is accessible through a prefix.
+	 */
+	synchronize_gk_instances(ip_prefix_fib, gk_conf);
+
+	/*
+	 * From now on, GK blocks must not have a reference
+	 * to @ip_prefix_fib.
+	 */
+
+	switch (ip_prefix_fib->action) {
+	case GK_FWD_GRANTOR: {
+		ret = del_gateway_from_neigh_table_locked(
+			ip_prefix, GK_FWD_GATEWAY_BACK_NET,
+			ip_prefix_fib->u.grantor.eth_cache, gk_conf);
+		if (ret < 0)
+			return -1;
```

Comment:
We must go to `initialize_fib_entry()` below before returning the error since the prefix was already removed from the LPM table. The same goes for the returns in the other `case` statements of this `switch`.